hy/compiler.py

```diff
@@ -59,31 +62,43 @@ def load_stdlib():
                 _stdlib[e] = module
 
 
-_compile_table = {}
-_decoratables = (ast.FunctionDef, ast.ClassDef)
-if PY35:
-    _decoratables += (ast.AsyncFunctionDef,)
-
-
 def ast_str(x, piecewise=False):
     if piecewise:
         return ".".join(ast_str(s) if s else "" for s in x.split("."))
     x = mangle(x)
     return x if PY3 else x.encode('UTF8')
 
 
-def builds(*types, **kwargs):
-    # A decorator that adds the decorated method to _compile_table for
-    # compiling `types`, but only if kwargs['iff'] (if provided) is
-    # true.
-    if not kwargs.get('iff', True):
-        return lambda fn: fn
+_special_form_compilers = {}
+_model_compilers = {}
+_decoratables = (ast.FunctionDef, ast.ClassDef)
+if PY35:
+    _decoratables += (ast.AsyncFunctionDef,)
+# _bad_roots are illegal at the top level.
+_bad_roots = tuple(ast_str(x) for x in (
+    "unquote", "unquote-splice", "unpack-mapping", "except"))
```

Comment:
Would it make sense to add `return`, `yield`, `yield-from` and `await` here?

Top level, these forms are currently accepted by the `hy` compiler, but generate invalid AST and thus causes Python's `compile` to crash with "SyntaxError: '...' outside function", which is completely detached from the source code.

Maybe it makes for a better user experience to catch it early.

Reply:
The comment is perhaps misleading. The detection for `_bad_roots` doesn't distinguish between forms inside a function and forms at the top level; it just triggers for any attempt to use one of the given names as a special form. `except` is illegal at the top level in the sense that it isn't a true special form; it's handled inside `@special("try", …)` instead.

Reply:
I've made the comment more descriptive.